ProjFS.Mac/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -653,3 +667,112 @@ VirtualizationRootHandle ActiveProvider_FindForPath(const char* _Nonnull path)
     
     return matchingHandle;
 }
+
+bool VirtualizationRoots_ProcessMayAccessOfflineRoots(pid_t pid)
+{
+    bool result = false;
+    
+    RWLock_AcquireShared(s_virtualizationRootsLock);
+    {
+        for (uint32_t i = 0; i < s_offlineIOPIDCount; ++i)
+        {
+            if (s_offlineIOPIDs[i] == pid)
+            {
+                result = true;
+                break;
+            }
+        }
+    }
+    RWLock_ReleaseShared(s_virtualizationRootsLock);
+
+    return result;
+}
+
+bool VirtualizationRoots_AddOfflineIOProcess(pid_t pid)
+{
+    bool success = false;
+    bool retry;
+    pid_t* resizeMemory = nullptr;
+    uint32_t resizeMemoryArrayLength = 0;
+    
+    do
+    {
+        retry = false;
+        RWLock_AcquireExclusive(s_virtualizationRootsLock);
```

Comment:
It would be easier for me to read if we did

if (s_offlineIOPIDCount >= s_maxOfflineIOPIDs)
{
   - resize logic
}

Add offlineRoot

Not sure why this needs to be in a do/while

Reply:
This is the usual thing where we need to avoid allocating memory while holding the lock, and when we drop the lock to allocate, we introduce the possibility of a race condition, hence the need for a retry. So the loop needs to be there. It doesn't have to be structured in exactly this way, but the other option takes the following form:

    Lock();
    while (needsResizing?)
    {
      // [work out new array size]
      Unlock();
      // [alloc memory for new array size]
      Lock();
      // [try to resize array with memory alloc'd during dropped lock]
    }
    // insert item
    Unlock();

So you end up with those unbalanced lock/unlock calls which definitely violate our coding guidelines; I don't think there's much difference readability wise, but I can flesh out the alternate version for a full comparison if you like.

Reply:
The block you have above is super readable to me.  Especially if you put that block in comments.

Something like:

```
Lock()
bool needsResizing = s_offlineIOPIDCount >= s_maxOfflineIOPIDs;
while (needsResizing)
{
   // We may need to retry since we attempt to allocate memory outside of the lock
  // [work out new array size]
  Unlock();
  // [alloc memory for new array size]
  Lock();
  // [try to resize array with memory alloc'd during dropped lock]
  // If we successful resized, needsResizing = false
}
// insert item
Unlock();

```
Perhaps some of the variables are also confusing me a bit.  resizeMemory sounds like a bool, but it's the resizeMemory.

This is a someone foreign paradigm to most folks on my team, but I'm sure you have extensive experience with it.  Any changes we do to make it ridiculously easy to understand will be appreciated.

Reply:
I've reworked the resize in a3618a4 - I'll leave this as a separate commit for now so the 2 variants can be compared side-by-side. I agree the new version is clearer.

Reply:
I really like the new verison.  Thank you for taking the time to simplify.